File: js/tests/test_csv.ts
import { Csv } from "zippedcsv";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";

describe("Csv", () => {
  test("construct from 2D array", () => {
    const csv = new Csv([["a", "b", "c"], ["1", "2", "3"]]);
    expect(csv.getHeader()).toEqual(["a", "b", "c"]);
    expect(csv.totalRows()).toBe(1);
    expect(csv.getRow(0)).toEqual({ a: "1", b: "2", c: "3" });
  });

  test("construct from array of objects", () => {
    const csv = new Csv([{ a: "1", b: "2" }, { a: "3", b: "4" }]);
    expect(csv.getHeader()).toEqual(["a", "b"]);
    expect(csv.totalRows()).toBe(2);
    expect(csv.getRow(1)).toEqual({ a: "3", b: "4" });
  });

  test("empty constructor", () => {
    const csv = new Csv();
    expect(csv.getHeader()).toEqual([]);
    expect(csv.getData()).toEqual([]);
    expect(csv.totalRows()).toBe(0);
    expect(csv.totalColumns()).toBe(0);
  });

  test("addRow with array", () => {
    const csv = new Csv([["a", "b", "c"]]);
    const result = csv.addRow(["1", "2", "3"]);
    expect(result).toBe(Csv.CSV_OK);
    expect(csv.totalRows()).toBe(1);
    expect(csv.getRow(0)).toEqual({ a: "1", b: "2", c: "3" });
  });

  test("addRow with object", () => {
    const csv = new Csv([["a", "b", "c"]]);
    const result = csv.addRow({ a: "1", b: "2", c: "3" });
    expect(result).toBe(Csv.CSV_OK);
    expect(csv.getRow(0)).toEqual({ a: "1", b: "2", c: "3" });
  });

  test("addRow size mismatch", () => {
    const csv = new Csv([["a", "b"]]);
    expect(csv.addRow(["1"])).toBe(Csv.CSV_ROW_SIZE_MISMATCH);
    expect(csv.addRow(["1", "2", "3"])).toBe(Csv.CSV_ROW_SIZE_MISMATCH);
  });

  test("addRow different keys", () => {
    const csv = new Csv([["a", "b"]]);
    expect(csv.addRow({ x: "1", y: "2" })).toBe(Csv.CSV_DIFFERENT_KEYS);
  });

  test("addColumn", () => {
    const csv = new Csv([["a", "b"], ["1", "2"], ["3", "4"]]);
    const result = csv.addColumn("c", ["5", "6"]);
    expect(result).toBe(Csv.CSV_OK);
    expect(csv.getHeader()).toEqual(["a", "b", "c"]);
    expect(csv.getColumn("c")).toEqual(["5", "6"]);
  });

  test("addColumn size mismatch", () => {
    const csv = new Csv([["a", "b"], ["1", "2"]]);
    expect(csv.addColumn("c", ["5", "6", "7"])).toBe(Csv.CSV_COLUMN_SIZE_MISMATCH);
  });

  test("removeRow", () => {
    const csv = new Csv([["a", "b"], ["1", "2"], ["3", "4"]]);
    expect(csv.removeRow(0)).toBe(Csv.CSV_OK);
    expect(csv.totalRows()).toBe(1);
    expect(csv.getRow(0)).toEqual({ a: "3", b: "4" });
  });

  test("removeRow out of bounds", () => {
    const csv = new Csv([["a", "b"], ["1", "2"]]);
    expect(csv.removeRow(5)).toBe(Csv.CSV_ROW_NOT_FOUND);
    expect(csv.removeRow(-1)).toBe(Csv.CSV_ROW_NOT_FOUND);
  });

  test("removeColumn", () => {
    const csv = new Csv([["a", "b", "c"], ["1", "2", "3"]]);
    expect(csv.removeColumn("b")).toBe(Csv.CSV_OK);
    expect(csv.getHeader()).toEqual(["a", "c"]);
    expect(csv.getRow(0)).toEqual({ a: "1", c: "3" });
  });

  test("removeColumn not found", () => {
    const csv = new Csv([["a", "b"]]);
    expect(csv.removeColumn("z")).toBe(Csv.CSV_COLUMN_NOT_FOUND);
  });

  test("setCellValue", () => {
    const csv = new Csv([["a", "b"], ["1", "2"]]);
    csv.setCellValue("b", 0, "99");
    expect(csv.getCell("b", 0)).toBe("99");
  });

  test("getColumn", () => {
    const csv = new Csv([["a", "b"], ["1", "2"], ["3", "4"]]);
    expect(csv.getColumn("a")).toEqual(["1", "3"]);
  });

  test("getData returns copies", () => {
    const csv = new Csv([["a"], ["1"]]);
    const data = csv.getData();
    data[0]["a"] = "mutated";
    expect(csv.getRow(0)["a"]).toBe("1");
  });

  test("loadFromString empty", () => {
    const csv = new Csv();
    expect(csv.loadFromString("")).toBe(Csv.CSV_OK);
    expect(csv.getHeader()).toEqual([]);
    expect(csv.getData()).toEqual([]);
  });

  test("loadFromString with header", () => {
    const csv = new Csv();
    csv.loadFromString("a,b,c\n1,2,3\n4,5,6", true, ",");
    expect(csv.getHeader()).toEqual(["a", "b", "c"]);
    expect(csv.totalRows()).toBe(2);
    expect(csv.getRow(0)).toEqual({ a: "1", b: "2", c: "3" });
  });

  test("loadFromString with semicolon delimiter", () => {
    const csv = new Csv();
    csv.loadFromString("a;b\n1;2", true, ";");
    expect(csv.getHeader()).toEqual(["a", "b"]);
    expect(csv.getRow(0)).toEqual({ a: "1", b: "2" });
  });

  test("toString round-trips data", () => {
    const original = "a,b,c\n1,2,3\n4,5,6";
    const csv = new Csv();
    csv.loadFromString(original, true, ",");
    const out = csv.toString(true, ",");
    const csv2 = new Csv();
    csv2.loadFromString(out, true, ",");
    expect(csv2.getData()).toEqual(csv.getData());
  });

  test("save and load from file", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = path.join(tmpDir, "test.csv");

    const csv = new Csv([["x", "y"], ["10", "20"]]);
    expect(csv.save(filePath, true, ",")).toBe(Csv.CSV_OK);

    const loaded = new Csv();
    expect(loaded.load(filePath, true, ",")).toBe(Csv.CSV_OK);
    expect(loaded.getHeader()).toEqual(["x", "y"]);
    expect(loaded.getRow(0)).toEqual({ x: "10", y: "20" });

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("load non-existent file", () => {
    const csv = new Csv();
    expect(csv.load("/non/existent/path.csv")).toBe(Csv.CSV_FILE_NOT_READABLE);
  });
});

File: cpp/tests/test_csv.cpp
#include <catch2/catch_all.hpp>

#include <filesystem>
#include <string>

#include "csv.h"

using namespace mc;

// ── Default constructor ────────────────────────────────────────────────────
TEST_CASE("Default constructor creates empty table", "[csv]") {
  Csv csv;
  CHECK(csv.TotalRows() == 0);
  CHECK(csv.TotalColumns() == 0);
  CHECK(csv.GetHeader().empty());
}

// ── 2-D array constructor ─────────────────────────────────────────────────
TEST_CASE("2-D array constructor sets header and rows", "[csv]") {
  Csv csv({{"Name", "Age"}, {"Alice", "30"}, {"Bob", "25"}});
  CHECK(csv.TotalColumns() == 2);
  CHECK(csv.TotalRows() == 2);
  CHECK(csv.GetHeader()[0] == "Name");
  CHECK(csv.GetCell("Age", 1) == "25");
}

TEST_CASE("2-D array constructor throws on mismatched row width", "[csv]") {
  CHECK_THROWS_AS((Csv({{"A", "B"}, {"only_one"}})),
                  std::invalid_argument);
}

// ── AddRow (vector) ────────────────────────────────────────────────────────
TEST_CASE("AddRow vector appends a row", "[csv]") {
  Csv csv({{"A", "B"}, {"1", "2"}});
  CHECK(csv.AddRow(std::vector<std::string>{"3", "4"}) == Csv::kOk);
  CHECK(csv.TotalRows() == 2);
  CHECK(csv.GetCell("B", 1) == "4");
}

TEST_CASE("AddRow vector returns kRowSizeMismatch on wrong size", "[csv]") {
  Csv csv({{"A", "B"}, {"1", "2"}});
  CHECK(csv.AddRow(std::vector<std::string>{"3"}) == Csv::kRowSizeMismatch);
}

// ── AddRow (map) ───────────────────────────────────────────────────────────
TEST_CASE("AddRow map appends a row by column name", "[csv]") {
  Csv csv({{"X", "Y"}, {"a", "b"}});
  CHECK(csv.AddRow(std::map<std::string,std::string>{{"X", "c"}, {"Y", "d"}}) == Csv::kOk);
  CHECK(csv.GetCell("X", 1) == "c");
}

TEST_CASE("AddRow map returns kDifferentKeys on missing key", "[csv]") {
  Csv csv({{"X", "Y"}, {"a", "b"}});
  CHECK(csv.AddRow(std::map<std::string,std::string>{{"X", "c"}, {"Z", "d"}}) == Csv::kDifferentKeys);
}

// ── AddColumn ─────────────────────────────────────────────────────────────
TEST_CASE("AddColumn appends a column", "[csv]") {
  Csv csv({{"A"}, {"1"}, {"2"}});
  CHECK(csv.AddColumn("B", {"x", "y"}) == Csv::kOk);
  CHECK(csv.TotalColumns() == 2);
  CHECK(csv.GetCell("B", 0) == "x");
}

TEST_CASE("AddColumn returns kColumnSizeMismatch on wrong size", "[csv]") {
  Csv csv({{"A"}, {"1"}});
  CHECK(csv.AddColumn("B", {"x", "y"}) == Csv::kColumnSizeMismatch);
}

// ── RemoveRow ─────────────────────────────────────────────────────────────
TEST_CASE("RemoveRow removes the correct row", "[csv]") {
  Csv csv({{"A"}, {"1"}, {"2"}, {"3"}});
  CHECK(csv.RemoveRow(1) == Csv::kOk);
  CHECK(csv.TotalRows() == 2);
  CHECK(csv.GetCell("A", 1) == "3");
}

TEST_CASE("RemoveRow returns kRowNotFound on bad index", "[csv]") {
  Csv csv({{"A"}, {"1"}});
  CHECK(csv.RemoveRow(5) == Csv::kRowNotFound);
  CHECK(csv.RemoveRow(-1) == Csv::kRowNotFound);
}

// ── RemoveColumn ──────────────────────────────────────────────────────────
TEST_CASE("RemoveColumn removes the correct column", "[csv]") {
  Csv csv({{"A", "B"}, {"1", "2"}});
  CHECK(csv.RemoveColumn("A") == Csv::kOk);
  CHECK(csv.TotalColumns() == 1);
  CHECK(csv.GetHeader()[0] == "B");
}

TEST_CASE("RemoveColumn returns kColumnNotFound on unknown name", "[csv]") {
  Csv csv({{"A"}, {"1"}});
  CHECK(csv.RemoveColumn("Z") == Csv::kColumnNotFound);
}

// ── SetCellValue ──────────────────────────────────────────────────────────
TEST_CASE("SetCellValue updates a cell", "[csv]") {
  Csv csv({{"V"}, {"old"}});
  csv.SetCellValue("V", 0, "new");
  CHECK(csv.GetCell("V", 0) == "new");
}

// ── GetColumn ─────────────────────────────────────────────────────────────
TEST_CASE("GetColumn returns all values for a column", "[csv]") {
  Csv csv({{"C"}, {"x"}, {"y"}, {"z"}});
  auto col = csv.GetColumn("C");
  REQUIRE(col.size() == 3);
  CHECK(col[2] == "z");
}

TEST_CASE("GetColumn throws on unknown column", "[csv]") {
  Csv csv({{"C"}, {"x"}});
  CHECK_THROWS_AS(csv.GetColumn("Z"), std::invalid_argument);
}

// ── GetRow ────────────────────────────────────────────────────────────────
TEST_CASE("GetRow returns a map of column->value", "[csv]") {
  Csv csv({{"A", "B"}, {"1", "2"}});
  auto row = csv.GetRow(0);
  CHECK(row["A"] == "1");
  CHECK(row["B"] == "2");
}

TEST_CASE("GetRow throws on bad index", "[csv]") {
  Csv csv({{"A"}, {"1"}});
  CHECK_THROWS_AS(csv.GetRow(99), std::out_of_range);
}

// ── LoadFromString / ToString round-trip ──────────────────────────────────
TEST_CASE("LoadFromString parses standard CSV", "[csv]") {
  Csv csv;
  int rc = csv.LoadFromString("Name;Age\nAlice;30\nBob;25\n");
  REQUIRE(rc == Csv::kOk);
  CHECK(csv.TotalRows() == 2);
  CHECK(csv.GetCell("Name", 0) == "Alice");
}

TEST_CASE("LoadFromString handles quoted fields with embedded delimiters",
          "[csv]") {
  Csv csv;
  csv.LoadFromString("A;B\n\"he;llo\";world\n");
  CHECK(csv.GetCell("A", 0) == "he;llo");
}

TEST_CASE("LoadFromString handles doubled-quote escaping", "[csv]") {
  Csv csv;
  csv.LoadFromString("A\n\"say \"\"hi\"\"\"\n");
  CHECK(csv.GetCell("A", 0) == "say \"hi\"");
}

TEST_CASE("ToString round-trip preserves data", "[csv]") {
  Csv orig({{"Name", "Score"}, {"Alice", "95"}, {"Bob", "80"}});
  std::string s = orig.ToString();
  Csv copy;
  copy.LoadFromString(s);
  CHECK(copy.TotalRows() == orig.TotalRows());
  CHECK(copy.GetCell("Score", 0) == "95");
}

TEST_CASE("LoadFromString comma-separated with comma delimiter", "[csv]") {
  Csv csv;
  csv.LoadFromString("A,B\n1,2\n", true, ',');
  CHECK(csv.GetCell("B", 0) == "2");
}

// ── Load / Save file ──────────────────────────────────────────────────────
TEST_CASE("Save writes a file and Load reads it back", "[csv]") {
  namespace fs = std::filesystem;
  std::string path = (fs::temp_directory_path() / "test_csv_roundtrip.csv").string();

  Csv orig({{"X", "Y"}, {"10", "20"}});
  REQUIRE(orig.Save(path) == Csv::kOk);

  Csv loaded;
  REQUIRE(loaded.Load(path) == Csv::kOk);
  CHECK(loaded.GetCell("Y", 0) == "20");

  fs::remove(path);
}

TEST_CASE("Load returns kFileNotReadable for missing file", "[csv]") {
  Csv csv;
  CHECK(csv.Load("/nonexistent/path/to/file.csv") == Csv::kFileNotReadable);
}
